bundles/org.eclipse.ui.ide/extensions/org/eclipse/ui/dialogs/FilteredResourcesSelectionDialog.java

```diff
@@ -676,93 +677,72 @@ public void applyStyles(TextStyle textStyle) {
 		private StyledString styleResourceExtensionMatch(IResource resource, String matchingString, Styler styler) {
 			StyledString str = new StyledString(resource.getName());
 			String resourceExtension = resource.getFileExtension();
-			int lastDotIndex = matchingString.lastIndexOf('.');
-			if (lastDotIndex == -1 || resourceExtension == null) {
+			if (resourceExtension == null) {
 				return str;
 			}
+			int matchingExtensionIndex = matchingString.lastIndexOf('.');
+			if (matchingExtensionIndex == -1) {
+				return str;
+			}
+
 			resourceExtension = '.' + resourceExtension;
-			int resourceExtensionIndex = resource.getName().lastIndexOf(resourceExtension);
-			String matchingExtension = matchingString.substring(lastDotIndex, matchingString.length());
+			int resourceExtensionIndex = resource.getName().length() - resourceExtension.length();
+			String matchingExtension = matchingString.substring(matchingExtensionIndex, matchingString.length());
 			for (Position markPosition : getMatchPositions(resourceExtension, matchingExtension)) {
 				str.setStyle(resourceExtensionIndex + markPosition.offset, markPosition.length, styler);
 			}
 			return str;
 		}
 
 		private List<Position> getMatchPositions(String string, String matching) {
-			final String originalMatching = matching;
 			List<Position> positions = new ArrayList<>();
 			if (matching.isEmpty() || string.isEmpty()) {
 				return positions;
 			}
 
+			// Pre-process the matching pattern
 			char lastChar = matching.charAt(matching.length() - 1);
-			if (lastChar == ' ') {
+			if (lastChar == ' ' || lastChar == '<') {
 				matching = matching.substring(0, matching.length() - 1);
-			} else if (lastChar == '<') {
-				matching = matching.substring(0, matching.length() - 2);
 			}
+
+			String originalTrimmedMatching = matching;
 			if (matching.indexOf('?') == -1 && matching.indexOf('*') == -1) {
+				// Expect that CamelCase search was possibly performed
+				// => convert pattern to "wildcards-like" one
```

Comment:
@vogella, after waking up, I've realized that we could probably make this even simpler before merging into master. Now we split the CamelCase pattern and join it by `*`, only to split it according to the very same character later on. So what about leaving just the 1st splitting step... I would like to look at it this or tomorrow evening (CET) if I have time.

Reply:
Done, see the force-push, hopefully correct. :)